jep/0000/README.adoc

```diff
@@ -0,0 +1,329 @@
+
+* Configuration UI
+* Extension Points for External Logging implementations
+* Base classes for external logging,
+which simplify implementations
+* Other shared utility classes
+
+== Specification
+
+=== Event class
+
+Event is an atomic entry of the external logging.
+In the data model it includes the following fields:
+
+* `id` (long) - ID of the message.
+  It must be unique within a single `LoggingMethod` writer
+* `message`(string) - Contains the message body.
+  It may be a single-line or a multi-line message.
+  The message may also contain annotations.
+* `timestamp` (long) - Timestamp of the message,
+  measured in milliseconds, between
+  the current time and midnight, January 1, 1970 UTC
+* `data` (`Map<String, Serializable>`) - additional metadata.
+  This metadata will be used to categorize event entries.
+
+Some metadata is mandatory.
+Requited fields are defined below.
+
+==== Event metadata
+
+The following metadata entries are required to be present
+for all event entries:
+
+* `jenkinsId` - Instance ID of the Jenkins instance
+* `loggableType`  - Type of the `Loggable` entry, e.g. `run`
+* `loggableId` - Identifier of the `Loggable` entry
+
+==== Run metadata
+
+The following metadata entries are required to be present
+for all `hudson.model.Run` event entries:
+
+* All event metadata
+* `jobId` - ID of the Job, should be provided by Unique ID Plugin
+
+==== Pipeline metadata
+
+* All Run metadata
+* `stepId` - Identifier of the step ("node")
+
+=== Extension Points
+
+The plugin should offer the following extension points:
+
+* `ExternalLoggingMethodFactory` -
+produced `ExternalLoggingMethod` instances for Runs and, eventually, other objects.
+This extension point implements `Describable`
+and should be configurable via UI.
+* `ExternalLogBrowserFactory` -
+Same as above, but for `ExternalLogBrowser`
+
+=== ExternalLoggingMethodFactory and classes
+
+Produces `ExternalLoggingMethod` instances for Runs and, eventually, other objects.
+This extension point implements `Describable`
+and should be configurable via UI.
+
+ExternalLoggingMethod is a new extension point on the top of `LoggingMethod`:
+
+* It implements `LoggingMethod` interfaces
+* It provides a new `createWriter()` method,
+which produces a `ExternalLoggingEventWriter` class instance.
+** This method automatically injects the mandatory metadata for events
+** An abstract `_createWriter()` method is offered for implementation by downstream implementations
+* Being compared to `LoggingMethod`,
+the implementation is designed to always perform logging
+from the agent side
+
+ExternalLoggingEventWriter class:
+
+* The class is `Serializable`.
+It will be sent to the agent side.
+* The class offers the following abstract methods:
+** `void writeEvent(Event event) throws IOException`
+- writing event to the remote storage
+* The class also stores metadata, which may be injected into the events
+** The class stores a Map of Serializable metadata entries
```

Comment:
Similarly, consider a non-Java map of metadata that can be consumed by other sources, plus a smaller Java serialized grouping.

Reply:
This makes parsing easier for non-java consumers too